### ISSUE TITLE
Macro invocation with curly-braces as a statement

### ISSUE THREAD

**Issue by NalaGinrut** [Rust-GCC/gccrs#22]
This issue is related to [0378-expr-macros ](https://github.com/rust-lang/rfcs/blob/master/text/0378-expr-macros.md) 

- Treat all macro invocations with parentheses, (), or square brackets, [], as expressions.
- Curly braces or a semicolon following the invocation to invoke a macro as a statement

However, it seems that `macro_invoce!{x}` doesn't work for nightly rustc:
```rust
macro_rules! maybe_return { ($e:expr) => ($e); }

fn frob(x: i32) -> i32{
    maybe_return! {x}
    // should return -1 
    -1
}
```

The result is:

```rust
error[E0308]: mismatched types
 --> ma.rs:4:20
  |
4 |     maybe_return! {x}
  |                    ^- help: consider using a semicolon here
  |                    |
  |                    expected `()`, found `i32`
```

I may incorrectly understand the RFC description, anyway, I leave it here for a placeholder.

Our curent parser doesn't recogonize it too:
```rust
i32 frob(x : i32)
BlockExpr:
{
 outer attributes: none
 inner attributes: none
 statements: none
 final expression: 
ArithmeticOrLogicalExpr: MacroInvocation: maybe_return!{x} - 1
}
```

**Comment by SimplyTheOther** [comment]
The RFC actually seems to be a bit ambiguous on what it really means, but the name of the RFC (expr-macros) and the name of the issue for the RFC ("Implement parse macro!()/macro![] as expr) implies to me that the primary point of the RFC is to force parsing macro invocations with parentheses or square brackets as an expression. As such, macro invocations with curly brackets _can_ be parsed as an expression, but do not have to be. 

That's just my interpretation of it, but it seems like it probably works with how rustc implements it.

TL;DR:

- Treat all macro invocations with parentheses, (), or square brackets, [], as expressions. 
- If the macro invocation has curly brackets, {}, it may be parsed as a statement depending on the context.
- If the macro invocation has a semicolon at the end, it must be parsed as a statement (either via ExpressionStatement or MacroInvocationWithSemi)

**Comment by NalaGinrut** [comment]
@SimplyTheOther Yes, I realized that something in the RFC doesn't actually hold in the implementation, I'll try to point them out if I found them. At least for now, I think we should do what `rustc` does, and enhance them according to the ideas in RFC when the time is proper. After all, following the RFC could be one of our selling points.

cc @philberty

**Comment by philberty** [comment]
Nice detective work on that i didn't know the RFC had forward name resolution on macros that's pretty neat. I've found the documentation of macros kind of lacking in rust so in my own programs I've always avoided them

**Comment by philberty** [comment]
can you raise this in a separate issue @bjorn3? i would like to track it as a ticket for our kanban board :)

**Comment by philberty** [comment]
Closing this as the final issue is tracked under #1048
